bcl/src/bc_esp8266.c

```diff
@@ -0,0 +1,864 @@
+}
+
+static void _uart_event_handler(bc_uart_channel_t channel, bc_uart_event_t event, void *param)
+{
+    (void) channel;
+    bc_esp8266_t *self = (bc_esp8266_t*)param;
+
+    if (event == BC_UART_EVENT_ASYNC_READ_DATA && self->_state == BC_ESP8266_STATE_IDLE)
+    {
+        bc_scheduler_plan_relative(self->_task_id, 100);
+        self->_state = BC_ESP8266_STATE_RECEIVE;
+    }
+}
+
+void bc_esp8266_set_event_handler(bc_esp8266_t *self, void (*event_handler)(bc_esp8266_t *, bc_esp8266_event_t, void *), void *event_param)
+{
+    self->_event_handler = event_handler;
+    self->_event_param = event_param;
+}
+
+void bc_esp8266_set_station_mode(bc_esp8266_t *self, char *ssid, char *password)
+{
+    self->_config.mode = BC_ESP8266_CONFIG_MODE_STATION;
+    strncpy(self->_config.ssid, ssid, 63);
+    strncpy(self->_config.password, password, 63);
+}
+
+void bc_esp8266_set_sntp(bc_esp8266_t *self, int timezone)
+{
+    bc_esp8266_set_sntp_with_servers(self, timezone, NULL, NULL, NULL);
+}
+
+void bc_esp8266_set_sntp_with_servers(bc_esp8266_t *self, int timezone, char *sntp_server1, char *sntp_server2, char *sntp_server3)
+{
+    self->_config.sntp_enabled = 1;
+    self->_config.sntp_timezone = timezone;
+    if (sntp_server1 != NULL)
+    {
+        strncpy(self->_config.sntp_server1, sntp_server1, 127);
+    }
+    else
+    {
+        strcpy(self->_config.sntp_server1, "0.pool.ntp.org");
+    }
+    if (sntp_server2 != NULL)
+    {
+        strncpy(self->_config.sntp_server2, sntp_server2, 127);
+    }
+    else
+    {
+        strcpy(self->_config.sntp_server2, "1.pool.ntp.org");
+    }
+    if (sntp_server3 != NULL)
+    {
+        strncpy(self->_config.sntp_server3, sntp_server3, 127);
+    }
+    else
+    {
+        strcpy(self->_config.sntp_server3, "2.pool.ntp.org");
+    }
+}
+
+bool bc_esp8266_is_ready(bc_esp8266_t *self)
+{
+    if (self->_state == BC_ESP8266_STATE_READY || self->_state == BC_ESP8266_STATE_IDLE)
+    {
+        return true;
+    }
+
+    return false;
+}
+
+bool bc_esp8266_connect(bc_esp8266_t *self)
+{
+    if (self->_state != BC_ESP8266_STATE_DISCONNECTED ||
+        self->_config.ssid[0] == '\0' || self->_config.password[0] == '\0')
+    {
+        return false;
+    }
+
+    // Enable CH_PD
+    bc_gpio_set_output(BC_GPIO_P8, 1);
+
+    self->_state = BC_ESP8266_STATE_INITIALIZE;
+
+    bc_scheduler_plan_now(self->_task_id);
+
+    return true;
+}
+
+bool bc_esp8266_disconnect(bc_esp8266_t *self)
+{
+    // Disable CH_PD
+    bc_gpio_set_output(BC_GPIO_P8, 0);
+
+    self->_state = BC_ESP8266_STATE_DISCONNECTED;
+
+    return true;
+}
+
+bool bc_esp8266_socket_connect(bc_esp8266_t *self, const char *type, const char *host, uint16_t port)
+{
+    if (!bc_esp8266_is_ready(self) || host == NULL || port == 0 || strlen(host) == 0 ||
+        (strlen(host) + 15) > BC_ESP8266_TX_MAX_PACKET_SIZE)
+    {
+        return false;
+    }
+
+    char buffer[BC_ESP8266_TX_MAX_PACKET_SIZE];
```

Comment:
In general, this feels like too bug buffer stack - I suggest to make it static, if possible. Thanks. P.